src/backend/utils/adt/network.c

```diff
@@ -389,6 +416,415 @@ network_cmp(PG_FUNCTION_ARGS)
+ *
+ * +----------+---------------------+
+ * | 1 bit IP |   31 bits netmask   |     (up to 1 bit
+ * |  family  |     (truncated)     |   netmask omitted)
+ * +----------+---------------------+
+ *
+ * 64-bit architecture:
+ *
+ * The most interesting case: we have space to store all netmasked bits,
+ * followed by the netmask size, followed by 25 bits of the subnet. We lay the
+ * bits out carefully this way so that the entire value can be compared as an
+ * integer while still adhering to the inet/cidr sorting rules stated above.
+ *
+ * 25 bits is more than enough to store most subnets (all but /6 or smaller),
+ * so in the vast majority of cases these keys should hold enough information
+ * to save trips to the heap.
+ *
+ * +----------+-----------------------+--------------+--------------------+
+ * | 1 bit IP |    32 bits netmask    |    6 bits    |   25 bits subnet   |
+ * |  family  |        (full)         | netmask size |    (truncated)     |
+ * +----------+-----------------------+--------------+--------------------+
+ *
+ * IPv6
+ * ----
+ *
+ * 32-bit architecture:
+ *
+ * +----------+---------------------+
+ * | 1 bit IP |   31 bits netmask   |    (up to 97 bits
+ * |  family  |     (truncated)     |   netmask omitted)
+ * +----------+---------------------+
+ *
+ * 64-bit architecture:
+ *
+ * +----------+---------------------------------+
+ * | 1 bit IP |         63 bits netmask         |    (up to 65 bits
+ * |  family  |           (truncated)           |   netmask omitted)
+ * +----------+---------------------------------+
+ *
+ */
+static Datum
+network_abbrev_convert(Datum original, SortSupport ssup)
+{
+	network_sortsupport_state *uss = ssup->ssup_extra;
+	inet	   *authoritative = DatumGetInetPP(original);
+	Datum		res;
+	Datum		ipaddr_int,
+				netmask_int,
+				subnet_int;
+	char		datum_size_left,
+				datum_subnet_size;
+
+	/* TODO: Is this enough, or should the possibility of a third value be */
+	/* handled on branches below? */
+	Assert(ip_family(authoritative) == PGSQL_AF_INET ||
+		   ip_family(authoritative) == PGSQL_AF_INET6);
+
+	res = (Datum) 0;
+	if (ip_family(authoritative) == PGSQL_AF_INET6)
+	{
+		/* Shift a 1 over to the datum's most significant bit. */
+		res = ((Datum) 1) << SIZEOF_DATUM * BITS_PER_BYTE - ABBREV_BITS_INET_FAMILY;
+	}
+
+	/*
+	 * For IPv6, create an integer from either the first 4 or 8 bytes of the
+	 * IP (depending on whether we're on 32 or 64-bit architecture). For IPv4,
+	 * always use all 4 bytes.
+	 *
+	 * We're consuming an array of char, so make sure to byteswap on little
+	 * endian systems (an inet's IP array emulates big endian in that the
+	 * first byte is always the most significant).
+	 */
+	if (ip_family(authoritative) == PGSQL_AF_INET6)
+	{
+		ipaddr_int = *((Datum *) ip_addr(authoritative));
+		ipaddr_int = DatumBigEndianToNative(ipaddr_int);
+	}
+	else
+	{
+		uint32		ipaddr_int32 = *((uint32 *) ip_addr(authoritative));
+#ifndef WORDS_BIGENDIAN
```

Comment:
Don't like the raw WORDS_BIGENDIAN here -- this is generally supposed to appear in very low level code.

Reply:
Can you use DatumBigEndianToNative() instead, while zeroing out the garbage bits afterwards? Sounds like a way to reduce variance between ipv4 and ipv6 code. More variability and commonality analysis can be done here.

Reply:
Interesting point. It's relatively easy to rewrite this as:

``` c
	ipaddr_int = *((Datum *) ip_addr(authoritative));
	ipaddr_int = DatumBigEndianToNative(ipaddr_int);

#if SIZEOF_DATUM == 8
	/* In the case of IPv4 (4 bytes) on a 64-bit machine (8 bytes) we ingested
	 * an extra 4 zeroed bytes in the most significant positions. Shift them
	 * off to get the integer back to its appropriate value. */
	if (ip_family(authoritative) == PGSQL_AF_INET)
	{
		ipaddr_int >>= (SIZEOF_DATUM * BITS_PER_BYTE - ip_maxbits(authoritative));
		Assert(ipaddr_int < PG_UINT32_MAX);
	}
#endif
```

How do you feel about that? It's not bad, but I sort of feel like delving into knowingly invalid bytes is a little like playing with fire.

Reply:
I think you should `Assert(<= PG_UINT32_MAX)`, but otherwise looks good.

It's not known invalid -- it's just not what you're interested in. But neither is a zeroed byte.

Reply:
Oops, good catch on the `<=`! (quite weird the test suite didn't catch that one given that it does `255.255.255.255`). Thanks!